File: 2019/Python/day1.py
### =============================================
### [ ADVENT OF CODE ] (https://adventofcode.com)
### 2019 - Mina Pêcheux: Python version
### ---------------------------------------------
### Day 1: The Tyranny of the Rocket Equation
### =============================================

# [ Input parsing functions ]
# ---------------------------
def parse_input(data):
    '''Parses the incoming data into processable inputs.
    
    :param data: Provided problem data.
    :type data: str
    :return: Parsed data.
    :rtype: list(int)
    '''
    return [ int(x) for x in data.split('\n') if x != '' ]

# [ Computation functions ]
# -------------------------
### PART I
def compute_fuel(mass):
    '''Computes the required fuel for a module of given mass.
    
    :param mass: The mass of the module to compute the fuel consumption for.
    :type mass: int
    :return: Required amount of fuel.
    :rtype: int
    '''
    return (mass // 3) - 2

### PART II
def compute_total_fuel(mass):
    '''Computes the total required fuel for a module of given mass and the
    added fuel, and so on. It works recursively until the computed amount of
    fuel is zero or negative.
    
    :param mass: The mass of the module to compute the fuel consumption for.
    :type mass: int
    :return: Required amount of fuel.
    :rtype: int
    '''
    f = compute_fuel(mass)
    if f <= 0:
        return 0
    return f + compute_total_fuel(f)
    
# [ Base tests ]
# --------------
def make_tests():
    '''Performs tests on the provided examples to check the result of the
    computation functions is ok.'''
    ### PART I
    assert compute_fuel(12) == 2
    assert compute_fuel(14) == 2
    assert compute_fuel(1969) == 654
    assert compute_fuel(100756) == 33583
    ### PART II
    assert compute_total_fuel(14) == 2
    assert compute_total_fuel(1969) == 966
    assert compute_total_fuel(100756) == 50346

if __name__ == '__main__':
    # check function results on example cases
    make_tests()
    
    # get input data
    data_path = '../data/day1.txt'
    inputs = parse_input(open(data_path, 'r').read())
    
    ### PART I
    solution = sum([ compute_fuel(i) for i in inputs ])
    print('PART I: solution = {}'.format(solution))
    
    ### PART II
    solution = sum([ compute_total_fuel(i) for i in inputs ])
    print('PART II: solution = {}'.format(solution))

File: 2019/Javascript/day1.js
/*
 * ================================================
 * [ ADVENT OF CODE ] (https://adventofcode.com)
 * 2019 - Mina Pêcheux: Javascript (NodeJS) version
 * ------------------------------------------------
 * Day 1: The Tyranny of the Rocket Equation
 * ================================================
 */
'use strict'

const _ = require('lodash')
const fs = require('fs')
const expect = require('chai').expect

// [ Input parsing functions ]
// ---------------------------
/**
 * Parses the incoming data into processable inputs.
 * @param {string} data - Provided problem data.
 * @returns {array(int)} - Parsed data.
 */
const parseInput = (data) => {
  return _.map(_.filter(_.split(data, '\n'), (l) => l.length > 0), Number)
}

// [ Computation functions ]
// -------------------------
/* PART I */
/**
 * Computes the required fuel for a module of given mass.
 * @param {int} mass - The mass of the module to compute the fuel consumption for.
 * @returns {int} - Required amount of fuel.
 */
const computeFuel = (mass) => _.floor(mass / 3) - 2

/* PART II */
/** Computes the total required fuel for a module of given mass and the
 * added fuel, and so on. It works recursively until the computed amount of
 * fuel is zero or negative.
 * @param {int} mass - The mass of the module to compute the fuel consumption for.
 * @returns {int} - Required amount of fuel.
 */
const computeTotalFuel = (mass) => {
  const f = computeFuel(mass)
  return (f <= 0) ? 0 : f + computeTotalFuel(f)
}
    
// [ Base tests ]
// --------------
/**
 * Performs tests on the provided examples to check the result of the
 * computation functions is ok.
 */
const makeTests = () => {
  /* Part I */
  expect(computeFuel(12)).to.equal(2)
  expect(computeFuel(14)).to.equal(2)
  expect(computeFuel(1969)).to.equal(654)
  expect(computeFuel(100756)).to.equal(33583)
  /* Part II */
  expect(computeTotalFuel(14)).to.equal(2)
  expect(computeTotalFuel(1969)).to.equal(966)
  expect(computeTotalFuel(100756)).to.equal(50346)
}

(() => {
  // check function results on example cases
  makeTests()
  
  // get input data
  const data = fs.readFileSync('../data/day1.txt')
  const inputs = parseInput(data)
  
  // Part I
  const solution1 = _.sum(_.map(inputs, (i) => computeFuel(i)))
  console.log(`PART I: solution = ${solution1}`)
  
  // Part I
  const solution2 = _.sum(_.map(inputs, (i) => computeTotalFuel(i)))
  console.log(`PART II: solution = ${solution2}`)
})()
